src/cpp/common/completion_queue_cc.cc

```diff
@@ -50,6 +50,26 @@ void CompletionQueue::CompleteAvalanching() {
   }
 }
 
+CompletionQueue::NextStatus CompletionQueue::DoThenAsyncNextInternal(
+    const std::function<void(void)>& F, void** tag, bool* ok,
+    gpr_timespec deadline) {
+  grpc_prepare_cq_tls_cache(cq_);
```

Comment:
Can we make a small class to handle the TLS cache here... constructor calls grpc_prepare... destructor does the cleanup (and could be in a cc file out of line since it's sufficiently complex)